### ISSUE TITLE
Autocompletion with Gocode doesn't work

### ISSUE THREAD

**Issue by pierrre** [joefitzgerald/go-plus#321]
If gocode is automatically started by Atom, the auto completion doesn't work.
If I start Gocode manually (with `gocode -s`), it works.

go-plus informations:

```
Go: go1.5.1 linux/amd64 (@/home/pierre/.gimme/versions/go1.5.1.src/bin/go)
GOPATH: /home/pierre/Go
Cover Tool: /home/pierre/.gimme/versions/go1.5.1.src/pkg/tool/linux_amd64/cover
Vet Tool: /home/pierre/.gimme/versions/go1.5.1.src/pkg/tool/linux_amd64/vet
Format Tool: /home/pierre/Go/bin/goimports
Lint Tool: /home/pierre/Go/bin/golint
Gocode Tool: /home/pierre/Go/bin/gocode
Godef Tool: /home/pierre/Go/bin/godef
Gocode Status: Enabled
Oracle Tool: /home/pierre/Go/bin/oracle
Git: /usr/bin/git
PATH: /usr/local/sbin:/usr/local/bin:/usr/bin:/usr/lib/jvm/default/bin:/usr/bin/site_perl:/usr/bin/vendor_perl:/usr/bin/core_perl
Atom: 1.2.4 (linux x64 4.2.5-1-ARCH)
```

My `go env`:

```
GOARCH="amd64"
GOBIN=""
GOEXE=""
GOHOSTARCH="amd64"
GOHOSTOS="linux"
GOOS="linux"
GOPATH="/home/pierre/Go"
GORACE=""
GOROOT="/home/pierre/.gimme/versions/go1.5.1.src"
GOTOOLDIR="/home/pierre/.gimme/versions/go1.5.1.src/pkg/tool/linux_amd64"
GO15VENDOREXPERIMENT="1"
CC="gcc"
GOGCCFLAGS="-fPIC -m64 -pthread -fmessage-length=0"
CXX="g++"
CGO_ENABLED="1"
```

My gocode config (`gocode set`):

```
propose-builtins false
lib-path ""
autobuild true
force-debug-output "/home/pierre/gocode-debug.log"
package-lookup-mode "go"
```

The Gocode debug log, started by Atom:

```
015/12/02 21:12:18 rebuilding package 
2015/12/02 21:12:18 package import: github.com/pierrre/imageserver/testdata
2015/12/02 21:12:18 package object: /home/pierre/Go/pkg/linux_amd64/github.com/pierrre/imageserver/testdata.a
2015/12/02 21:12:18 package source dir: /home/pierre/Go/src/github.com/pierrre/imageserver/testdata
2015/12/02 21:12:18 package source files: []
2015/12/02 21:12:18 Autobuild error: exec: "go": executable file not found in $PATH
2015/12/02 21:12:18 Import path "github.com/pierrre/imageserver/testdata" was not resolved
2015/12/02 21:12:18 Gocode's build context is:
2015/12/02 21:12:18  GOROOT: /home/pierre/.gimme/versions/go1.5.1.src
2015/12/02 21:12:18  GOPATH: /home/pierre/Go
2015/12/02 21:12:18  GOOS: linux
2015/12/02 21:12:18  GOARCH: amd64
2015/12/02 21:12:18  GBProjectRoot: ""
2015/12/02 21:12:18  lib-path: ""
2015/12/02 21:12:18 extracted expression tokens: params
2015/12/02 21:12:18 Offset: 0
2015/12/02 21:12:18 Number of candidates found: 0
2015/12/02 21:12:18 Candidates are:
2015/12/02 21:12:18 =======================================================
```

Gocode env variables (`cat /proc/xxx/environ`), started by Atom:

```
XDG_VTNR=1XDG_SESSION_ID=c1GIO_LAUNCHED_DESKTOP_FILE_PID=7612DESKTOP_STARTUP_ID=gnome-shell-784-pierre-vm-/usr/bin/atom-3_TIME566759XDG_MENU_PREFIX=gnome-SHELL=/bin/zshGJS_DEBUG_OUTPUT=stderrGJS_DEBUG_TOPICS=JS ERROR;JS LOGUSER=pierreSSH_AUTH_SOCK=/run/user/1000/keyring/sshSESSION_MANAGER=local/pierre-vm:@/tmp/.ICE-unix/658,unix/pierre-vm:/tmp/.ICE-unix/658USERNAME=pierreGIO_LAUNCHED_DESKTOP_FILE=/usr/share/applications/atom.desktopMOZ_PLUGIN_PATH=/usr/lib/mozilla/pluginsDESKTOP_SESSION=gnomeMAIL=/var/spool/mail/pierrePATH=/usr/local/sbin:/usr/local/bin:/usr/bin:/usr/lib/jvm/default/bin:/usr/bin/site_perl:/usr/bin/vendor_perl:/usr/bin/core_perlHG=/usr/bin/hgPWD=/home/pierreXDG_SESSION_TYPE=x11LANG=fr_FR.UTF-8GDM_LANG=fr_FR.UTF-8GDMSESSION=gnomeSSH_ASKPASS=/usr/lib/seahorse/seahorse-ssh-askpassSHLVL=1HOME=/home/pierreXDG_SEAT=seat0GNOME_DESKTOP_SESSION_ID=this-is-deprecatedXDG_SESSION_DESKTOP=gnomeLOGNAME=pierreDBUS_SESSION_BUS_ADDRESS=unix:path=/run/user/1000/busWINDOWPATH=1DISPLAY=:0XDG_RUNTIME_DIR=/run/user/1000XDG_CURRENT_DESKTOP=GNOMEPYTHON=python2XAUTHORITY=/run/user/1000/gdm/Xauthority_=/usr/bin/nohupGOOGLE_API_KEY=AIzaSyAQfxPJiounkhOjODEO5ZieffeBv6yft2QATOM_HOME=/home/pierre/.atomNODE_PATH=/usr/share/atom/resources/app.asar/exportsNODE_ENV=productionGOPATH=/home/pierre/Go
```

Gocode env variables, started manually:

```
LANG=fr_FR.UTF-8DISPLAY=:0SHLVL=1LOGNAME=pierreXDG_VTNR=1PWD=/home/pierreHG=/usr/bin/hgMOZ_PLUGIN_PATH=/usr/lib/mozilla/pluginsXAUTHORITY=/run/user/1000/gdm/XauthorityXDG_SESSION_ID=c1DESKTOP_SESSION=gnomeXDG_SESSION_DESKTOP=gnomeGDMSESSION=gnomeGNOME_DESKTOP_SESSION_ID=this-is-deprecatedUSERNAME=pierreWINDOWPATH=1DBUS_SESSION_BUS_ADDRESS=unix:path=/run/user/1000/busVTE_VERSION=4201MAIL=/var/spool/mail/pierreGJS_DEBUG_TOPICS=JS ERROR;JS LOGXDG_MENU_PREFIX=gnome-GDM_LANG=fr_FR.UTF-8GJS_DEBUG_OUTPUT=stderrXDG_SESSION_TYPE=x11SHELL=/bin/zshWINDOWID=29360134TERM=xterm-256colorSESSION_MANAGER=local/pierre-vm:@/tmp/.ICE-unix/658,unix/pierre-vm:/tmp/.ICE-unix/658XDG_CURRENT_DESKTOP=GNOMESSH_AUTH_SOCK=/run/user/1000/keyring/sshPATH=/home/pierre/.gimme/versions/go1.5.1.src/bin:/usr/local/sbin:/usr/local/bin:/usr/bin:/usr/lib/jvm/default/bin:/usr/bin/site_perl:/usr/bin/vendor_perl:/usr/bin/core_perl:/home/pierre/Logiciels:/home/pierre/.gimme/bin:/home/pierre/Go/bin:/home/pierre/Logiciels/android-sdk/platform-toolsHOME=/home/pierreXDG_SEAT=seat0XDG_RUNTIME_DIR=/run/user/1000SSH_ASKPASS=/usr/lib/seahorse/seahorse-ssh-askpassUSER=pierreOLDPWD=/home/pierreZSH=/home/pierre/.oh-my-zshPAGER=lessLESS=-RLC_CTYPE=fr_FR.UTF-8LSCOLORS=GxfxcxdxbxegedabagacadEDITOR=nanoCDPATH=.:/home/pierre:/home/pierre/Go/src:/home/pierre/Go/src/github.com/pierrreGIMME_GO_VERSION=1.5.1GO15VENDOREXPERIMENT=1_GORC=/home/pierre/.gorc_GIMME=/home/pierre/.gimmeGIMME_TYPE=sourceGIMME_SILENT_ENV=1_GIMME_ENV=/home/pierre/.gimme/envs/go1.5.1.envGOPATH=/home/pierre/Go_=/home/pierre/Go/bin/gocode
```

Atom env variables:

```
XDG_VTNR=1XDG_SESSION_ID=c1SHELL=/bin/zshUSER=pierreUSERNAME=pierreMOZ_PLUGIN_PATH=/usr/lib/mozilla/pluginsPATH=/usr/local/sbin:/usr/local/bin:/usr/bin:/usr/lib/jvm/default/bin:/usr/bin/site_perl:/usr/bin/vendor_perl:/usr/bin/core_perlMAIL=/var/spool/mail/pierreDESKTOP_SESSION=gnomeHG=/usr/bin/hgXDG_SESSION_TYPE=x11PWD=/home/pierreLANG=fr_FR.UTF-8GDM_LANG=fr_FR.UTF-8GDMSESSION=gnomeSSH_ASKPASS=/usr/lib/seahorse/seahorse-ssh-askpassXDG_SEAT=seat0HOME=/home/pierreSHLVL=0LOGNAME=pierreXDG_SESSION_DESKTOP=gnomeDBUS_SESSION_BUS_ADDRESS=unix:path=/run/user/1000/busWINDOWPATH=1XDG_RUNTIME_DIR=/run/user/1000DISPLAY=:0XDG_CURRENT_DESKTOP=GNOMEXAUTHORITY=/run/user/1000/gdm/XauthorityGNOME_DESKTOP_SESSION_ID=this-is-deprecatedXDG_MENU_PREFIX=gnome-SESSION_MANAGER=local/pierre-vm:@/tmp/.ICE-unix/658,unix/pierre-vm:/tmp/.ICE-unix/658SSH_AUTH_SOCK=/run/user/1000/keyring/sshGIO_LAUNCHED_DESKTOP_FILE=/usr/share/applications/atom.desktopGIO_LAUNCHED_DESKTOP_FILE_PID=7612GJS_DEBUG_OUTPUT=stderrGJS_DEBUG_TOPICS=JS ERROR;JS LOGDESKTOP_STARTUP_ID=gnome-shell-784-pierre-vm-/usr/bin/atom-3_TIME566759
```

**Comment by pierrre** [comment]
By the way, if I start manually Atom from my terminal (`atom`), (gocode is started by Atom), it works.

**Comment by pierrre** [comment]
I think gocode fails to run go, because the $PATH is not set correctly by Atom (or go-plus?)

**Comment by andradei** [comment]
$PATH not set is precisely the issue. Starting atom from a terminal solves the problem for now.

**Comment by joefitzgerald** [comment]
@pierrre Is this still an issue for you?

**Comment by pierrre** [comment]
With the previous version of go-plus, I started Atom from my app menu.
It didn't use my `$GOPATH` defined in `.zshrc`,  but it used the parameter from go-plus config.
Go test/coverage was working, because it used the parameter.
Gocode was not working, because `$GOPATH` was not defined for Atom. (since it was started from the app menu)

With the new version of go-plus, a valide $GOPATH/$PATH is required for everything.
That's why I start Atom from the command line.
Everything works fine now.

I tried all the tricks from other Github issues, but it doesn't work for me.
I can't use Atom/go-plus if it's started from the app menu.

My config:
- Arch Linux
- Gnome Shell
- ZSH

Should I close the issue?

**Comment by joefitzgerald** [comment]
@pierrre Is zsh also your login shell? I think your .zshrc is only getting used for your terminal sessions, but not for your desktop session.

**Comment by andradei** [comment]
Is it true that Atom 1.7 will "fix" this issue? Meaning it will get environment variables without starting it from the terminal?

**Comment by joefitzgerald** [comment]
Atom 1.7 fixes it on OS X; no patching of the environment is done on Linux at this time. 1.7 also includes changes that allow different windows to have different environments, and also allows the environment for an existing window to be updated from the terminal.

/cc @lee-dohm

**Comment by liming84525** [comment]
after updating to atom beta it seems i don't have apm in zsh ..

**Comment by liming84525** [comment]
and it didn't resolve the issue

**Comment by joefitzgerald** [comment]
This will be resolved with the release of go-plus v5.0.0, occurring on Monday December 19th.
